docs/recipes/immutability-helpers.md

```diff
@@ -0,0 +1,190 @@
+# Immutability Helpers
+
+Redux is a tiny pattern that represents states as immutable objects. Redux was originally designed for React. Most Redux concepts, such as pure functions, are centered around the React ecosystem. Nowadays Redux is not directly related to React.
+
+The cornerstone of Redux is immutability. Immutabilty is an amazing pattern to minimise unpredictable behaviour in our code. We're not going to cover functional programming in this article. However we're going to look at very useful packages that are called "immutability helpers". Most developers have to deal with, so called, "deep objects" and most important follow the immutability concept, when it comes to changing the value of some deeply nested property.
+
+## immer
+
+`immer` is the most popular library that allows you to work with immutable objects in a more convenient way. Given the following code:
+
+```ts
+export interface Task {
+  title: string;
+  dates: {
+    startDate: string;
+    dueDate: string;
+  };
+}
+
+export interface TrelloStateModel {
+  tasks: {
+    [taskId: string]: Task;
+  };
+}
+
+@State<TrelloStateModel>({
+  name: 'trello',
+  defaults: {
+    tasks: {}
+  }
+})
+export class TrelloState {}
+```
+
+Let's imagine that we're faced with the task of changing the `dueDate` property:
+
+```ts
+export class UpdateDueDate {
+  static readonly type = '[Trello] Update due date';
+  constructor(public taskId: string, public dueDate: string) {}
+}
+```
+
+Let's see how we would implement the `updateDueDate` action handler:
+
+```ts
+export class TrelloState {
+
+  @Action(UpdateDueDate)
+  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
+    ctx.setState(state => ({
+      tasks: {
+        ...state.tasks,
+        [action.taskId]: {
+          ...state.tasks[action.taskId],
+          dates: {
+            ...state.tasks[action.taskId].dates,
+            dueDate: action.dueDate
+          }
+        }
+      }
+    }));
+  }
+
+}
+```
```

Comment:
This section above that sets out the example scenario and showcases the problem of doing a deep update should really be in its own section titled "The Problem".
Then below this would be a section called "Some Solutions" under which each of the tools are showcased.